### PR TITLE
Handle oversized clipboard for GUI protocol 1.8

### PR DESCRIPTION
**Pull request by alimirjamali**
If vmside clipboard is over the maximum limit, sending one byte over maximum limit should trigger inter-vm clipboard rejection instead of truncation.

Fixes: https://github.com/QubesOS/qubes-issues/issues/9296
Fixes: https://github.com/QubesOS/qubes-issues/issues/5220

### CODE REVIEW

File: gui-agent/vmside.c
Comment (marmarek) [inline gui-agent/vmside.c:988]: This value should also take the protocol version into account.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: gui-agent/vmside.c
Comment (marmarek) [reply gui-agent/vmside.c:988]: No, currently it doesn't - with your change in gui-common, the `MAX_CLIPBOARD_SIZE` will be 256k, so it won't enter this block for values above  65k but below 256k. But as just commented in the gui-common PR, better to keep `MAX_CLIPBOARD_SIZE` unchanged to avoid issues like this, and introduce new name for the new limit.

File: rpm_spec/gui-agent.spec.in
Comment (marmarek) [inline rpm_spec/gui-agent.spec.in:74]: Leave this to me after merge. This needs to be done only after the other version is committed (and preferably built, otherwise CI fails).

**Review by marmarek — CHANGES_REQUESTED** [review]

File: rpm_spec/gui-agent.spec.in
Comment (alimirjamali) [reply rpm_spec/gui-agent.spec.in:74]: Sure. Will do it in one hour. I am finalizing PR for issue 9029 (`Ability to hide (a) qube(s) from updates`) at the moment. And will come back to this shortly.

Comment (alimirjamali) [reply rpm_spec/gui-agent.spec.in:74]: I converted the status of this PR to draft (as well its related PRs). This needs much more work and testing.

Comment (alimirjamali) [reply rpm_spec/gui-agent.spec.in:74]: This is reverted to previous situation

File: gui-agent/vmside.c
Comment (alimirjamali) [reply gui-agent/vmside.c:988]: This is now fixed. The value of `MAX_CLIPBOARD_SIZE` remains the previous value in GUI Protocol V1.7

The new values for Protocol v1.8 are new definitions within `qubes-gui-common`

**Comment by alimirjamali** [comment]
@marmarek 
For handling INCR requests (clipboard sizes over 256KiB), I could not think of anything better than handling it at the vmside. And sending a warning message (that is not currently implemented) instead of a meaningless 1 byte.  The code is the 2nd commit to this PR.

**Comment by marmarek** [comment]
PipelineRetryFailed

**Comment by alimirjamali** [comment]
> PipelineRetryFailed

The pipeline fails again. Most probably because merging the `qui-common` commit alone is not enough for the builder? And it as to be included in the latest release?

https://gitlab.com/QubesOS/qubes-gui-agent-linux/-/jobs/8144115759#L3094

**Comment by marmarek** [comment]
CI still uses R4.2 here. Can you add a commit that switches it to R4.3 here? (just s/4.2/4.3/ in the .gitlab-ci.yml)

**Comment by alimirjamali** [comment]
> CI still uses R4.2 here. Can you add a commit that switches it to R4.3 here? (just s/4.2/4.3/ in the .gitlab-ci.yml)

Done. For both `gui-agent` and `gui-daemon`

**Comment by alimirjamali** [comment]
> CI still uses R4.2 here. Can you add a commit that switches it to R4.3 here? (just s/4.2/4.3/ in the .gitlab-ci.yml)

It is better now. checks, build and install unittest are green. The repro tests still fail. Is there anything else I can do here?

**Comment by marmarek** [comment]
reprotest fails with:
```
12:54:50 [qb.gitlab-reprotestv2] vmside.c:990:33: error: ...QUBES_GUID_MIN_CLIPBOARD_4X... undeclared (first use in this function)
```
looks like it still uses old gui-common... let me check

**Comment by marmarek** [comment]
PipelineRetry

**Comment by alimirjamali** [comment]
PipelineRetryFailed

**Comment by marmarek** [comment]
PipelineRetryFailed

File: gui-agent/vmside.c
Comment (marmarek) [inline gui-agent/vmside.c:1065]: Test says there is extra `\x00` copied:
```suggestion
                                sizeof(INCR_WARNING)-1, g->protocol_version);
```

Comment (alimirjamali) [reply gui-agent/vmside.c:1065]: I will do it.

One other possibility that comes to mind would be sending `char something[MAX_CLIPBOARD_BUFFER_SIZE+1] = {0}` to trigger the xside oversized detection mechanism. A little bit wasteful but it will work.

**Comment by marmarek** [comment]
Some commits are mixed up - the CI one has bumping protocol version, and 1d4a71771a5233f115e06ddb796046d95ac9045e has a fix to message length.
Furthermore, I think the final newline in the INCR message isn't needed.

**Comment by alimirjamali** [comment]
> Some commits are mixed up - the CI one has bumping protocol version, and [1d4a717](https://github.com/QubesOS/qubes-gui-agent-linux/commit/1d4a71771a5233f115e06ddb796046d95ac9045e) has a fix to message length. Furthermore, I think the final newline in the INCR message isn't needed.

Could I squash them all together to one commit? This one is not as long as `gui-daemon` one.

**Comment by marmarek** [comment]
CI switch should be IMO in separate commit, but others can be squashed together.

**Comment by alimirjamali** [comment]
> CI switch should be IMO in separate commit, but others can be squashed together.

Is it good now?

**Comment by marmarek** [comment]
Yes, now looks good :)

**Review by marmarek — APPROVED** [review]

**Pull request merged**
